File: frontend/src/api/frequencyAdvisor.ts
/**
 * Wash frequency recommendation logic.
 * Ported from backend/wardrobe/frequency_advisor.py so the APK is self-contained.
 */

import type { FrequencyAdvice, LaundryConstraints, RiskLevel, WardrobeItemForPlan } from "./types";

const DAY_MS = 24 * 60 * 60 * 1000;

const WEAR_COUNT_THRESHOLDS: Record<string, number> = {
  underwear: 1,
  sock: 1,
  sport: 1,
  sports: 1,
  towel: 3,
  "t-shirt": 2,
  tee: 2,
  tshirt: 2,
  shirt: 2,
  hoodie: 3,
  sweater: 4,
  wool: 4,
  jeans: 5,
  denim: 5,
  pants: 4,
  jacket: 8,
  coat: 8,
  bedding: 2,
  sheet: 2,
  sheets: 2,
  duvet: 2,
  内衣: 1,
  贴身: 1,
  袜: 1,
  运动: 1,
  速干: 1,
  浴巾: 3,
  毛巾: 3,
  t恤: 2,
  "t 恤": 2,
  衬衫: 2,
  卫衣: 3,
  羊毛: 4,
  毛衣: 4,
  牛仔: 5,
  裤: 4,
  夹克: 8,
  外套: 8,
  床单: 2,
  被套: 2,
};

const DAY_CYCLE_THRESHOLDS: Record<string, number> = {
  underwear: 1,
  sock: 1,
  sport: 1,
  sports: 1,
  towel: 3,
  "t-shirt": 2,
  tee: 2,
  tshirt: 2,
  shirt: 2,
  hoodie: 7,
  sweater: 14,
  wool: 14,
  jeans: 7,
  denim: 7,
  pants: 7,
  jacket: 30,
  coat: 30,
  bedding: 14,
  sheet: 14,
  sheets: 14,
  duvet: 14,
  内衣: 1,
  贴身: 1,
  袜: 1,
  运动: 1,
  速干: 1,
  浴巾: 3,
  毛巾: 3,
  t恤: 2,
  "t 恤": 2,
  衬衫: 2,
  卫衣: 7,
  羊毛: 14,
  毛衣: 14,
  牛仔: 7,
  裤: 7,
  夹克: 30,
  外套: 30,
  床单: 14,
  被套: 14,
};

const DAY_CYCLE_TERMS = new Set([
  "towel", "浴巾", "毛巾", "hoodie", "sweater", "wool", "jeans", "denim", "pants", "jacket", "coat",
  "bedding", "sheet", "sheets", "duvet", "卫衣", "羊毛", "毛衣", "牛仔", "裤", "夹克", "外套", "床单", "被套",
]);
const LOW_FREQUENCY_TERMS = new Set(["jeans", "denim", "sweater", "wool", "jacket", "coat", "牛仔", "羊毛", "毛衣", "夹克", "外套"]);
const SPORT_TERMS = new Set(["sport", "sports", "运动", "速干", "sweat", "出汗"]);
const STAIN_TERMS = new Set(["stain", "污渍", "油渍"]);
const HYGIENE_TERMS = new Set(["underwear", "sock", "内衣", "贴身", "袜"]);
const FREQUENCY_RISK_KEYS = new Set(["shrink", "color_bleed", "deform", "pilling", "dryer_damage"]);

export function adviseFrequency(item: WardrobeItemForPlan, constraints: LaundryConstraints): FrequencyAdvice {
  const text = searchText(item);
  const wearThreshold = wearThresholdFor(text);
  const dayThreshold = dayThresholdFor(text);
  const wearCount = nonNegativeInteger(item.wear_count_since_wash);
  const urgentItemIds = new Set(constraints.urgent_item_ids.map((id) => id.trim()).filter(Boolean));
  const dayInfo = usageDayInfo(item);
  const dayDue = dayInfo != null && dayInfo.days >= dayThreshold;
  const wearDue = wearCount >= wearThreshold;
  const sportRelated = containsAny(text, SPORT_TERMS);
  const hygieneRelated = constraints.hygiene_sensitive && containsAny(text, HYGIENE_TERMS);
  const immediateUseBased = sportRelated || hygieneRelated;
  const cycleSensitive = containsAny(text, DAY_CYCLE_TERMS) || containsAny(text, LOW_FREQUENCY_TERMS);
  const reasons: string[] = [];
  let score = 0;

  if (dayInfo) {
    if (dayDue) {
      score += 45;
      reasons.push(`距离${dayInfo.basis}已 ${dayInfo.days} 天，达到建议清洗周期 ${dayThreshold} 天。`);
    } else {
      reasons.push(`距离${dayInfo.basis} ${dayInfo.days} 天，未达到建议清洗周期 ${dayThreshold} 天。`);
    }
  } else if (immediateUseBased && wearDue) {
    reasons.push(`贴身、运动或出汗类衣物已使用 ${wearCount} 次，建议及时清洗。`);
  } else if (cycleSensitive) {
    reasons.push(`这类衣物更适合按清洗周期判断，暂无上次洗涤或穿用日期；已使用 ${wearCount} 次仅作参考。`);
  } else if (wearDue) {
    reasons.push(`已穿 ${wearCount} 次，达到建议穿着上限 ${wearThreshold} 次。`);
  } else {
    reasons.push(`已穿 ${wearCount} 次，未达到建议穿着上限 ${wearThreshold} 次。`);
  }

  if (wearDue) {
    if (immediateUseBased) {
      score += 45;
    } else if (dayInfo || cycleSensitive) {
      score += 20;
    } else {
      score += 45;
    }
    if (dayInfo) {
      reasons.push(`已穿 ${wearCount} 次，达到建议穿着上限 ${wearThreshold} 次。`);
    }
  }

  if (urgentItemIds.has(item.profile.item_id)) {
    score += 25;
    reasons.push("该衣物被标记为本次急用，优先级提高。");
  }

  if (sportRelated) {
    score += 35;
    reasons.push("运动后或出汗后穿着，建议及时清洗。");
  }

  if (containsAny(text, STAIN_TERMS)) {
    score += 35;
    reasons.push("用户记录有明显污渍，建议本次优先处理。");
  }

  if (containsAny(text, LOW_FREQUENCY_TERMS) && !dayDue && wearCount < wearThreshold) {
    score -= 15;
    reasons.push("牛仔、羊毛或外套类衣物可适当少洗，减少褪色、缩水和变形。");
  }

  const penalty = riskPenalty(item);
  if (penalty) {
    score -= penalty;
    reasons.push("历史或抽取结果提示存在缩水、掉色、变形等风险，频率建议不会强行要求机洗。");
  }

  if (hygieneRelated) {
    score += 20;
    reasons.push("贴身或高卫生敏感衣物，建议提高换洗频率。");
  }

  score = Math.max(0, Math.min(score, 100));

  return {
    item_id: item.profile.item_id,
    priority_score: score,
    recommendation: recommendationText(score),
    reasons: dedupe(reasons),
  };
}

export function adviseAllFrequencies(
  items: WardrobeItemForPlan[],
  constraints: LaundryConstraints,
): FrequencyAdvice[] {
  validateUniqueItemIds(items);
  return items
    .map((item) => adviseFrequency(item, constraints))
    .sort((a, b) => b.priority_score - a.priority_score);
}

/** Return item ids whose score meets or exceeds minScore. */
export function recommendedItemIds(
  items: WardrobeItemForPlan[],
  constraints: LaundryConstraints,
  minScore: number,
): string[] {
  const safeMinScore = Number.isFinite(minScore) ? minScore : 0;
  return adviseAllFrequencies(items, constraints)
    .filter((advice) => advice.priority_score >= safeMinScore)
    .map((advice) => advice.item_id);
}

// ─── helpers ────────────────────────────────────────────────────────────

function wearThresholdFor(text: string): number {
  const matches: number[] = [];
  for (const [term, threshold] of Object.entries(WEAR_COUNT_THRESHOLDS)) {
    if (termMatches(text, term)) matches.push(threshold);
  }
  if (!matches.length) return 4; // default threshold for unknown items
  return Math.min(...matches);
}

function dayThresholdFor(text: string): number {
  const matches: number[] = [];
  for (const [term, threshold] of Object.entries(DAY_CYCLE_THRESHOLDS)) {
    if (termMatches(text, term)) matches.push(threshold);
  }
  if (!matches.length) return 7;
  return Math.min(...matches);
}

function usageDayInfo(item: WardrobeItemForPlan): { days: number; basis: string } | null {
  const firstWornAt = timestamp(item.first_worn_after_wash_at);
  if (firstWornAt != null) {
    return { days: elapsedDays(firstWornAt), basis: "本轮首次穿用" };
  }
  const lastWashedAt = timestamp(item.last_washed_at);
  if (lastWashedAt != null) {
    return { days: elapsedDays(lastWashedAt), basis: "上次洗涤" };
  }
  return null;
}

function timestamp(value: unknown): number | null {
  if (typeof value !== "string" || !value.trim()) return null;
  const time = Date.parse(value);
  return Number.isFinite(time) ? time : null;
}

function elapsedDays(startTime: number): number {
  return Math.max(0, Math.floor((Date.now() - startTime) / DAY_MS));
}

function nonNegativeInteger(value: number): number {
  return Number.isFinite(value) && Number.isInteger(value) && value >= 0 ? value : 0;
}

function termMatches(text: string, term: string): boolean {
  if (/^[a-z0-9 -]+$/i.test(term)) {
    return new RegExp(`(^|[^a-z0-9])${escapeRegExp(term)}([^a-z0-9]|$)`).test(text);
  }
  return text.includes(term);
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function riskPenalty(item: WardrobeItemForPlan): number {
  const relevant = Object.entries(item.profile.risks)
    .filter(([key]) => FREQUENCY_RISK_KEYS.has(key))
    .map(([, level]) => level);
  if (relevant.some((l) => l === "high")) return 15;
  if (relevant.some((l) => l === "medium")) return 8;
  const notes = item.user_notes.join(" ");
  if (["缩水", "掉色", "起球", "shrink", "bleed", "pill"].some((term) => notes.includes(term))) return 8;
  return 0;
}

function recommendationText(score: number): string {
  if (score >= 75) return "建议本次优先清洗";
  if (score >= 45) return "建议本次清洗";
  if (score >= 25) return "可视时间和机器情况决定是否清洗";
  return "可暂缓清洗";
}

function searchText(item: WardrobeItemForPlan): string {
  const p = item.profile;
  return [
    p.name,
    p.user_note,
    ...Object.keys(p.material_ratios),
    ...p.colors,
    ...p.care_forbidden,
    ...item.user_notes,
  ]
    .join(" ")
    .toLowerCase();
}

function containsAny(text: string, terms: Set<string>): boolean {
  for (const term of terms) {
    if (termMatches(text, term)) return true;
  }
  return false;
}

function dedupe(items: string[]): string[] {
  return [...new Set(items)];
}

function validateUniqueItemIds(items: WardrobeItemForPlan[]): void {
  const seen = new Set<string>();
  const duplicates: string[] = [];

  for (const item of items) {
    const itemId = item.profile.item_id;
    if (seen.has(itemId)) {
      duplicates.push(itemId);
    } else {
      seen.add(itemId);
    }
  }

  if (duplicates.length) {
    throw new Error(`items duplicate item_id: ${dedupe(duplicates).join(", ")}`);
  }
}

File: backend/wardrobe/frequency_advisor.py
"""Wash frequency recommendation implementation."""

from __future__ import annotations

import math

from backend.shared.models import FrequencyAdvice, LaundryConstraints, RiskLevel, WardrobeItem
from backend.shared.utils import contains_any, dedupe


_FREQUENCY_THRESHOLDS = {
    "underwear": 1,
    "sock": 1,
    "sport": 1,
    "sports": 1,
    "t-shirt": 2,
    "shirt": 2,
    "hoodie": 3,
    "sweater": 4,
    "wool": 4,
    "jeans": 5,
    "denim": 5,
    "pants": 4,
    "coat": 8,
    "bedding": 1,
    "内衣": 1,
    "贴身": 1,
    "袜": 1,
    "运动": 1,
    "速干": 1,
    "t恤": 2,
    "t 恤": 2,
    "衬衫": 2,
    "卫衣": 3,
    "羊毛": 4,
    "毛衣": 4,
    "牛仔": 5,
    "裤": 4,
    "外套": 8,
    "床单": 1,
    "被套": 1,
}

_LOW_FREQUENCY_TERMS = {"jeans", "denim", "sweater", "wool", "coat", "牛仔", "羊毛", "毛衣", "外套"}
_SPORT_TERMS = {"sport", "sports", "运动", "速干", "sweat", "出汗"}
_STAIN_TERMS = {"stain", "污渍", "油渍"}
_FREQUENCY_RISK_KEYS = {"shrink", "color_bleed", "deform", "pilling", "dryer_damage"}


def advise_frequency(
    item: WardrobeItem,
    constraints: LaundryConstraints,
) -> FrequencyAdvice:
    """Recommend whether and how urgently one item should be washed."""

    _validate_item(item)
    _validate_constraints(constraints)
    search_text = _search_text(item)
    threshold = _threshold_for(search_text)
    reasons: list[str] = []
    score = 0.0

    if item.wear_count_since_wash >= threshold:
        score += 45
        reasons.append(f"已穿 {item.wear_count_since_wash} 次，达到建议清洗阈值 {threshold} 次。")
    else:
        reasons.append(f"已穿 {item.wear_count_since_wash} 次，未达到建议清洗阈值 {threshold} 次。")

    if item.profile.item_id in constraints.urgent_item_ids:
        score += 25
        reasons.append("该衣物被标记为本次急用，优先级提高。")

    if contains_any(search_text, _SPORT_TERMS):
        score += 35
        reasons.append("运动后或出汗后穿着，建议及时清洗。")

    if contains_any(search_text, _STAIN_TERMS):
        score += 35
        reasons.append("用户记录有明显污渍，建议本次优先处理。")

    if contains_any(search_text, _LOW_FREQUENCY_TERMS) and item.wear_count_since_wash < threshold:
        score -= 15
        reasons.append("牛仔、羊毛或外套类衣物可适当少洗，减少褪色、缩水和变形。")

    penalty = _risk_penalty(item)
    if penalty:
        score -= penalty
        reasons.append("历史或抽取结果提示存在缩水、掉色、变形等风险，频率建议不会强行要求机洗。")

    if constraints.hygiene_sensitive and contains_any(search_text, {"underwear", "sock", "内衣", "贴身", "袜"}):
        score += 20
        reasons.append("贴身或高卫生敏感衣物，建议提高换洗频率。")

    score = max(0.0, min(score, 100.0))
    return FrequencyAdvice(
        item_id=item.profile.item_id,
        priority_score=score,
        recommendation=_recommendation(score),
        reasons=dedupe(reasons),
    )


def advise_all_frequencies(
    items: list[WardrobeItem],
    constraints: LaundryConstraints,
) -> list[FrequencyAdvice]:
    """Return frequency advice sorted by priority score, highest first."""

    _validate_items(items)
    _validate_constraints(constraints)
    return sorted(
        [advise_frequency(item, constraints) for item in items],
        key=lambda advice: advice.priority_score,
        reverse=True,
    )


def recommended_item_ids(
    items: list[WardrobeItem],
    constraints: LaundryConstraints,
    min_score: float,
) -> list[str]:
    """Return item ids whose explicit score meets min_score."""

    safe_min_score = _min_score(min_score)
    return [
        advice.item_id
        for advice in advise_all_frequencies(items, constraints)
        if advice.priority_score >= safe_min_score
    ]


def _threshold_for(text: str) -> int:
    matches = [
        threshold
        for term, threshold in _FREQUENCY_THRESHOLDS.items()
        if term in text
    ]
    if not matches:
        raise ValueError("cannot infer wash-frequency threshold from wardrobe item data")
    return min(matches)


def _validate_item(value: object) -> None:
    if not isinstance(value, WardrobeItem):
        raise ValueError("item must be a WardrobeItem")
    _non_negative_int(value.wear_count_since_wash, "wear_count_since_wash")


def _validate_items(value: object) -> None:
    if not isinstance(value, list):
        raise ValueError("items must be a list")
    for index, item in enumerate(value):
        if not isinstance(item, WardrobeItem):
            raise ValueError(f"items[{index}] must be a WardrobeItem")


def _validate_constraints(value: object) -> None:
    if not isinstance(value, LaundryConstraints):
        raise ValueError("constraints must be LaundryConstraints")
    _item_id_list(value.urgent_item_ids, "urgent_item_ids")
    _boolean(value.hygiene_sensitive, "hygiene_sensitive")


def _min_score(value: object) -> float:
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise ValueError("min_score must be numeric")
    score = float(value)
    return score if math.isfinite(score) else 0.0


def _item_id_list(value: object, field_name: str) -> None:
    if not isinstance(value, list):
        raise ValueError(f"{field_name} must be a list of non-empty strings")
    if not all(isinstance(item, str) and item.strip() for item in value):
        raise ValueError(f"{field_name} must be a list of non-empty strings")


def _boolean(value: object, field_name: str) -> None:
    if not isinstance(value, bool):
        raise ValueError(f"{field_name} must be a boolean")


def _non_negative_int(value: object, field_name: str) -> None:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{field_name} must be a non-negative integer")
    if value < 0:
        raise ValueError(f"{field_name} must be a non-negative integer")


def _search_text(item: WardrobeItem) -> str:
    profile = item.profile
    return " ".join(
        [
            profile.name,
            profile.user_note,
            " ".join(profile.material_ratios.keys()),
            " ".join(profile.colors),
            " ".join(profile.care_forbidden),
            " ".join(profile.source_notes),
            " ".join(item.user_notes),
        ]
    ).lower()


def _risk_penalty(item: WardrobeItem) -> float:
    relevant = {
        key: level
        for key, level in item.profile.risks.items()
        if key in _FREQUENCY_RISK_KEYS
    }
    if any(level == RiskLevel.HIGH for level in relevant.values()):
        return 15.0
    if any(level == RiskLevel.MEDIUM for level in relevant.values()):
        return 8.0
    notes = " ".join(item.user_notes)
    if any(term in notes for term in ("缩水", "掉色", "起球", "shrink", "bleed", "pill")):
        return 8.0
    return 0.0


def _recommendation(score: float) -> str:
    if score >= 75:
        return "建议本次优先清洗"
    if score >= 45:
        return "建议本次清洗"
    if score >= 25:
        return "可视时间和机器情况决定是否清洗"
    return "可暂缓清洗"

